canopen/pdo.py

```diff
@@ -15,48 +12,97 @@
+    if direction_lower in ["transmit", "tx"]:
+        PDO = TPDO
+    elif direction_lower in ["receive", "rx"]:
+        PDO = RPDO
+    else:
+        logger.critical("Direction %s is not supported" % direction)
+        return maps
+    for idx in range(0x200):
+        com_index = com_offset + idx
+        map_index = map_offset + idx
+        if com_index in pdo_node.node.object_dictionary:
+            com_entry = pdo_node.node.object_dictionary[com_index]
+            if com_entry.subindices[1].raw:
+                cob_id = com_entry.subindices[1].raw
+            else:
+                logger.warning("Don't know how to handle communication "
+                               "index 0x{:X}".format(com_index))
+                continue
+            maps[com_index] = PDO(cob_id, pdo_node, com_index, map_index)
+    return maps
 
-class PdoNode(collections.Mapping):
-    """Represents a slave unit."""
 
+class RemotePdoBase(collections.Mapping):
+    """Represents a slave unit."""
     def __init__(self, node):
         self.network = None
         self.node = node
-        self.rx = Maps(0x1400, 0x1600, self)
-        self.tx = Maps(0x1800, 0x1A00, self)
+        self.subscriptions = set()
+        self.rx = {}
+        self.tx = {}
 
     def __iter__(self):
-        for pdo_maps in (self.rx, self.tx):
-            for pdo_map in pdo_maps.values():
-                for var in pdo_map.map:
-                    yield var.name
+        return chain(iter(sorted(self.rx.keys())),
+                     iter(sorted(self.tx.keys())))
 
     def __getitem__(self, key):
-        for pdo_maps in (self.rx, self.tx):
-            for pdo_map in pdo_maps.values():
-                for var in pdo_map.map:
-                    if var.name == key:
-                        return var
+        if key in self.rx:
+            return self.rx[key]
+        if key in self.tx:
+            return self.tx[key]
         raise KeyError("%s was not found in any map" % key)
 
     def __len__(self):
-        count = 0
-        for pdo_maps in (self.rx, self.tx):
-            for pdo_map in pdo_maps.values():
-                count += len(pdo_map)
-        return count
-
-    def read(self):
-        """Read PDO configuration from node using SDO."""
-        for pdo_maps in (self.rx, self.tx):
-            for pdo_map in pdo_maps.values():
-                pdo_map.read()
-
-    def save(self):
-        """Save PDO configuration to node using SDO."""
-        for pdo_maps in (self.rx, self.tx):
-            for pdo_map in pdo_maps.values():
-                pdo_map.save()
+        return len(self.rx) + len(self.tx)
+
+    def stop(self):
+        pass
+
+    def setup(self):
+        for pdos in (self.rx, self.tx):
+            for pdo in pdos.values():
+                pdo.setup()
+
+    def cleanup(self):
+        for pdos in (self.rx, self.tx):
+            for pdo in pdos.values():
+                pdo.cleanup()
+
+
+class RemotePdoNode(RemotePdoBase):
```

Comment:
Should also implement the `stop` method.

Reply:
The question is what the implication of `stop` would be? In my conception a remote PDO node is a silent listener, which does no sending at all, thus nothing is to stop and the base case is taken.

Reply:
The RemotePdoNode sends RPDOs to the node. The `stop` method stops the RPDOs.